File: app/src/main/java/jp/juggler/subwaytooter/api/entity/CustomEmoji.java
package jp.juggler.subwaytooter.api.entity;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;

import jp.juggler.subwaytooter.util.LogCategory;
import jp.juggler.subwaytooter.util.Utils;

public class CustomEmoji {
	static final LogCategory log = new LogCategory( "CustomEmoji" );
	
	// shortcode (コロンを含まない)
	@NonNull final public String shortcode;
	
	// 画像URL
	@NonNull final public String url;
	
	// アニメーションなしの画像URL
	@Nullable final public String static_url;
	
	private CustomEmoji( @NonNull String shortcode, @NonNull String url, @Nullable String static_url ){
		this.shortcode = shortcode;
		this.url = url;
		this.static_url = static_url;
	}
	
	public static CustomEmoji parse( JSONObject src ){
		if( src == null ) return null;
		String shortcode = Utils.optStringX( src, "shortcode" );
		String url = Utils.optStringX( src, "url" );
		String static_url = Utils.optStringX( src, "static_url" ); // may null
		if( TextUtils.isEmpty( shortcode ) || TextUtils.isEmpty( url ) ){
			return null;
		}
		return new CustomEmoji( shortcode, url, static_url );
	}
	
	public static class List extends ArrayList< CustomEmoji > {
	}
	
	public static List parseList( JSONArray src, @NonNull String instance ){
		if( src == null ) return null;
		List dst = new List();
		for( int i = 0, ie = src.length() ; i < ie ; ++ i ){
			CustomEmoji item = parse( src.optJSONObject( i ) );
			if( item != null ) dst.add( item );
		}
		Collections.sort( dst, new Comparator< CustomEmoji >() {
			@Override public int compare( CustomEmoji a, CustomEmoji b ){
				return a.shortcode.compareToIgnoreCase( b.shortcode );
			}
		} );
		log.d( "parseList: parse %d emojis for %s.", dst.size(), instance );
		return dst;
	}
	
	public static class Map extends HashMap< String, CustomEmoji > {
		// キー： shortcode (コロンを含まない)
	}
	
	public static Map parseMap( JSONArray src, @NonNull String instance ){
		if( src == null ) return null;
		Map dst = new Map();
		for( int i = 0, ie = src.length() ; i < ie ; ++ i ){
			CustomEmoji item = parse( src.optJSONObject( i ) );
			if( item != null ) dst.put( item.shortcode, item );
		}
		log.d( "parseMap: parse %d emojis for %s.", dst.size(), instance );
		return dst;
	}
	
}

File: app/src/main/java/jp/juggler/subwaytooter/api/entity/CustomEmoji.kt
package jp.juggler.subwaytooter.api.entity

import jp.juggler.emoji.EmojiBase
import jp.juggler.util.JsonArray
import jp.juggler.util.JsonObject
import jp.juggler.util.notEmpty

class CustomEmoji(
	val apDomain: Host,
	val shortcode: String, // shortcode (コロンを含まない)
	val url: String, // 画像URL
	val static_url: String?, // アニメーションなしの画像URL
	val aliases: ArrayList<String>? = null,
	val alias: String? = null,
	val visible_in_picker: Boolean = true,
	val category: String? = null
) : EmojiBase(), Mappable<String> {

    fun makeAlias(alias: String) = CustomEmoji(
		apDomain = apDomain,
		shortcode = shortcode,
		url = url,
		static_url = static_url,
		alias = alias
	)

    override val mapKey: String
        get() = shortcode

    companion object {

        val decode: (Host, JsonObject) -> CustomEmoji = { apDomain, src ->
            CustomEmoji(
				apDomain = apDomain,
				shortcode = src.stringOrThrow("shortcode"),
				url = src.stringOrThrow("url"),
				static_url = src.string("static_url"),
				visible_in_picker = src.optBoolean("visible_in_picker", true),
				category = src.string("category")
			)
        }

        val decodeMisskey: (Host, JsonObject) -> CustomEmoji = { apDomain, src ->
            val url = src.string("url") ?: error("missing url")

            CustomEmoji(
				apDomain = apDomain,
				shortcode = src.string("name") ?: error("missing name"),
				url = url,
				static_url = url,
				aliases = parseAliases(src.jsonArray("aliases"))
			)
        }

        private fun parseAliases(src: JsonArray?): ArrayList<String>? {
            var dst = null as ArrayList<String>?
            if (src != null) {
                val size = src.size
                if (size > 0) {
                    dst = ArrayList(size)
                    src.forEach {
                        val str = it?.toString()?.notEmpty()
                        if (str != null) dst.add(str)
                    }
                }
            }
            return if (dst?.isNotEmpty() == true) dst else null
        }
    }

}
